torch/utils/flop_counter.py

```diff
@@ -0,0 +1,306 @@
+    # NB(chilli): Should be 2 * c_in - 1 technically for FLOPs.
+    flop = batch_size * prod(conv_shape) * c_out * prod(dims) * 2 * c_in
+    return flop
+
+def conv_flop(input_shapes: List[Any], output_shapes: List[Any]) -> int:
+    """
+    Count flops for convolution.
+    """
+    x_shape, w_shape = input_shapes[:2]
+    out_shape = output_shapes[0]
+    transposed = input_shapes[6]
+
+    return conv_flop_count(x_shape, w_shape, out_shape, transposed=transposed)
+
+def transpose_shape(shape):
+    return [shape[1], shape[0]] + list(shape[2:])
+
+def conv_backward_flop(input_shapes: List[Any], output_shapes: List[Any]) -> int:
+    grad_out_shape, x_shape, w_shape = input_shapes[:3]
+    output_mask = input_shapes[-1]
+    fwd_transposed = input_shapes[7]
+    flop_count = 0
+
+    if output_mask[0]:
+        grad_input_shape = get_shape(output_shapes[0])
+        flop_count += conv_flop_count(grad_out_shape, w_shape, grad_input_shape, not fwd_transposed)
+    if output_mask[1]:
+        grad_weight_shape = get_shape(output_shapes[1])
+        flop_count += conv_flop_count(transpose_shape(x_shape), grad_out_shape, grad_weight_shape, fwd_transposed)
+
+    return flop_count
+
+
+flop_mapping = {
+    aten.mm: mm_flop,
+    aten.addmm: addmm_flop,
+    aten.bmm: bmm_flop,
+    aten.baddbmm: baddbmm_flop,
+    aten.convolution: conv_flop,
+    aten._convolution: conv_flop,
+    aten.convolution_backward: conv_backward_flop,
+}
+
+def normalize_tuple(x):
+    if not isinstance(x, tuple):
+        return (x,)
+    return x
+
+
+# Thanks BingChat!
+def convert_num_to_suffix_str(number):
+    # Define the suffixes for different orders of magnitude
+    suffixes = ["", "K", "M", "B", "T"]
+    # Find the index of the appropriate suffix based on the number of digits
+    index = max(0, min(len(suffixes) - 1, (len(str(number)) - 1) // 3))
+    # Divide the number by 1000^index and format it to two decimal places
+    value = "{:.2f}".format(number / (1000 ** index))
+    # Return the value and the suffix as a string
+    return value + suffixes[index]
+
+class FlopCounterMode(TorchDispatchMode):
+    """
+    ``FlopCounterMode`` is a context manager that counts the number of
+    flops within its context. It does this using a ``TorchDispatchMode``.
+
+    It also supports hierarchical output by passing a module to FlopCounterMode on construction.
+
+    Example usage
+
+    .. code-block:: python
+
+        mod = ...
+        flop_counter = FlopCounterMode(mod)
+        with flop_counter:
+            mod.sum().backward()
+
+    """
+    def __init__(self, mod: torch.nn.Module = None, depth: int = 2, display: bool = True, custom_mapping: Dict[Any, Any] = None):
+        self.flop_counts: Dict[Any, Any] = defaultdict(lambda: defaultdict(int))
+        self.depth = depth
+        self.parents = ["Global"]
+        self.display = display
+        if custom_mapping is None:
+            custom_mapping = {}
+        if mod is not None:
```

Comment:
You don't actually need the user to pass in the module or use custom Function.
You can use global hooks like `torch.nn.modules.module.register_module_forward_pre_hook()` that will be called for ALL modules.

More generally, you can use `forward_pre`/`forward` hook to do the forward push-pop and the `full_backward_pre_hook`/`full_backward_hook` to do the backward push pop.

Reply:
I think that will be a good API improvement. But in that case, what shall we do if users call two different models (which might have shared or distinct layers with the same name) in the context? This might get tricky to handle.

Reply:
oh, full_backward_pre_hook is a new API :P

Reply:
In general, I don't think global module hooks make that much sense given that we're trying to register a hierarchy. Concretely, the hook doesn't tell us where in the module hierarchy it lives.

Reply:
Hmm... I needed to do the "clone" thing in order to work around this issu
```
RuntimeError: Output 0 of BackwardHookFunctionBackward is a view and is being modified inplace. This view was created inside a custom Function (or because an input was returned as-is) and the autograd logic to handle view+inplace would override the custom backward associated with the custom Function, leading to incorrect gradients. This behavior is forbidden. You can fix this by cloning the output of the custom Function.
```

How would I do this with `full_backward_pre_hook`?

Reply:
> Concretely, the hook doesn't tell us where in the module hierarchy it lives.

But you can use a global stack to track where you are in the hierarchy right now. Like you know what the current Module is and when you enter a new one, make the current its parent and set the new one as the new current.

> Hmm... I needed to do the "clone" thing in order to work around this issu

I'm not sure what you mean by the clone thing. But yes if you have custom Function that is an issue.
If you have time, I think we can rework this to use multi-Tensor hooks instead of custom Function and that will remove this constraint :)

Reply:
If I have a module like `ReLU` where the forwards pass is
```
def forward(self, x):
    return x.relu_()
```

And I install a backwards hook, it throws that error.